File: client/python/hakesclient/extensions/mongodb.py
# Copyright 2024 The HAKES Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from ..components.store import Store
from pymongo import MongoClient, UpdateOne
from typing import List, Tuple
import struct


class MongoDB(Store):
    """
    MongoDB store implementation for HAKES.
    This class extends the Store interface to provide MongoDB-specific functionality.
    """

    def _validate_mongodb_address(self, addr: str) -> str:
        """
        Validate and format the MongoDB address.

        Args:
            addr (str): The address to validate.

        Returns:
            str: The validated and formatted MongoDB address.

        Raises:
            ValueError: If the address is empty or invalid.
        """
        if not addr:
            raise ValueError("MongoDB address is empty")

        if not addr.startswith("mongodb://"):
            addr = "mongodb://" + addr

        return addr

    def __init__(
        self, addr: str, db_name: str = "hakes", collection_name: str = "default"
    ):
        """
        Initialize the MongoDB store with the address of the MongoDB server.

        Args:
            addr (str): The address of the MongoDB server.
            db_name (str): The database name to use.
            collection_name (str): The collection name to use.
        """
        super().__init__(addr)
        self.addr = self._validate_mongodb_address(addr)
        self.client = MongoClient(self.addr)
        self.db = self.client[db_name]
        self.collection = self.db[collection_name]
        # Ensure index on key for fast lookup
        self.collection.create_index("key", unique=True)
        # Counter collection for xid auto-increment
        self.counters = self.db["counters"]
        # Initialize counter if not present
        if self.counters.find_one({"_id": collection_name}) is None:
            self.counters.insert_one({"_id": collection_name, "seq": 0})
        self._counter_key = collection_name

    def connected(self) -> bool:
        """
        Check if the MongoDB connection is established.
        """
        return self.client.admin.command("ismaster")["ok"] == 1

    def connect(self):
        """
        Connect to the MongoDB server (no-op, handled by MongoClient).
        """
        pass

    def disconnect(self):
        """
        Disconnect from the MongoDB server.
        """
        if hasattr(self, "client"):
            self.client.close()

    # deletor to call disconnect
    def __del__(self):
        """
        Destructor to ensure the MongoDB connection is closed when the object is deleted.
        """
        try:
            self.disconnect()
        except Exception:
            pass

    def _get_next_xids(self, batch_size: int) -> list[bytes]:
        # Atomically increment the counter by batch_size and get the starting value
        doc = self.counters.find_one_and_update(
            {"_id": self._counter_key},
            {"$inc": {"seq": batch_size}},
            return_document=True,
            upsert=True,
        )
        # The new value of seq is the last reserved id, so calculate the range
        last_id = doc["seq"]
        first_id = last_id - batch_size + 1
        # always encode to 8 bytes, big-endian signed
        return [struct.pack(">q", i) for i in range(first_id, last_id + 1)]

    def put(
        self, keys: List[str], value: List[bytes], xids: List[bytes] | None
    ) -> Tuple[bool, List[bytes]]:
        """
        Put key-value pairs into the store. Auto-increment xid if not provided, and batch the operation.
        XIDs are always stored as 8-byte (64-bit signed) values.
        """
        if len(keys) != len(value):
            raise ValueError("keys and value must have the same length")
        if xids and len(xids) != len(keys):
            raise ValueError("xids and keys must have the same length")

        try:
            n = len(keys)
            # Determine which indices need auto xids
            if xids is None:
                xids = self._get_next_xids(n)
            docs = []

            for i, key in enumerate(keys):
                doc = {"key": key, "value": value[i], "xid": xids[i]}
                docs.append(doc)
            # Prepare bulk upsert operations
            operations = [
                UpdateOne({"key": doc["key"]}, {"$set": doc}, upsert=True)
                for doc in docs
            ]
            self.collection.bulk_write(operations, ordered=False)
            return True, xids
        except Exception as e:
            return False, []

    def get_by_keys(self, keys: List[str]) -> Tuple[List[bytes], List[bytes]]:
        """
        Get values and xids by a list of keys from the store, using a batched query.
        Returns two lists: values and xids, in the same order as input keys.
        If a key is missing, returns b'' for value and b'' for xid.
        """
        if not keys:
            return [], []
        docs = self.collection.find({"key": {"$in": keys}})
        key_to_doc = {doc["key"]: doc for doc in docs if "key" in doc}
        values = [key_to_doc.get(key, {}).get("value", b"") for key in keys]
        xids = [key_to_doc.get(key, {}).get("xid", b"") for key in keys]
        return values, xids

    def get_by_ids(self, xids: List[bytes]) -> list[bytes]:
        """
        Get values by a list of xids from the store, using a batched query.
        Preserves the order of input xids in the output.
        """
        if not xids:
            return []
        # Convert xids to a list of unique values for the query
        xids_set = list(set(xids))
        # Query all at once
        docs = self.collection.find({"xid": {"$in": xids_set}})
        # Build a mapping from xid to value
        xid_to_value = {
            doc["xid"]: doc["value"] for doc in docs if "xid" in doc and "value" in doc
        }
        # Return results in the same order as input xids
        results = [xid_to_value.get(xid, b"") for xid in xids]
        return results

    def delete(self, key: str) -> bool:
        """
        Delete a key-value pair from the store.
        """
        result = self.collection.delete_one({"key": key})
        return result.deleted_count > 0


# https://www.mongodb.com/resources/products/platform/mongodb-auto-increment
# https://www.mongodb.com/resources/languages/python

File: hakes-server/store/mongodb/mongodb.cpp
/*
 * Copyright 2024 The HAKES Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "mongodb.h"

#include <mongocxx/client.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/uri.hpp>
#include <bsoncxx/builder/stream/document.hpp>
#include <bsoncxx/builder/stream/array.hpp>
#include <bsoncxx/document/view.hpp>
#include <bsoncxx/types.hpp>
#include <stdexcept>
#include <algorithm>
#include <cstring>
#include <cstdint>

namespace hakes {

using bsoncxx::builder::stream::document;
using bsoncxx::builder::stream::array;
using bsoncxx::builder::stream::finalize;
using bsoncxx::builder::stream::open_document;
using bsoncxx::builder::stream::close_document;

// Global MongoDB instance (must be created once per process)
static mongocxx::instance inst{};

std::string MongoDB::ValidateMongoDBAddress(const std::string& addr) {
  if (addr.empty()) {
    throw std::invalid_argument("MongoDB address is empty");
  }

  if (addr.find("mongodb://") == 0 || addr.find("mongodb+srv://") == 0) {
    return addr;
  }

  return "mongodb://" + addr;
}

MongoDB::MongoDB(const std::string& addr,
                 const std::string& db_name,
                 const std::string& collection_name)
    : Store(ValidateMongoDBAddress(addr)),
      counter_key_(collection_name) {
  try {
    // Create MongoDB client
    client_ = std::make_shared<mongocxx::client>(mongocxx::uri{addr_});

    // Get database and collection
    db_ = client_->database(db_name);
    collection_ = db_.collection(collection_name);
    counters_ = db_.collection("counters");

    // Create unique index on key for fast lookup
    auto index_opts = mongocxx::options::index{};
    collection_.create_index(document{} << "key" << 1 << finalize, index_opts);

    // Initialize counter if not present
    auto counter_doc = counters_.find_one(
        document{} << "_id" << counter_key_ << finalize);

    if (!counter_doc) {
      counters_.insert_one(
          document{} << "_id" << counter_key_ << "seq" << int64_t(0) << finalize);
    }
  } catch (const std::exception& e) {
    throw std::runtime_error(std::string("Failed to connect to MongoDB: ") + e.what());
  }
}

MongoDB::~MongoDB() {
  Disconnect();
}

bool MongoDB::Connected() {
  try {
    if (!client_) return false;
    auto admin = client_->database("admin");
    auto result = admin.run_command(document{} << "ismaster" << 1 << finalize);
    // If we get here, we're connected
    return true;
  } catch (...) {
    return false;
  }
}

void MongoDB::Connect() {
  // No-op, handled by constructor
}

void MongoDB::Disconnect() {
  client_.reset();
}

std::vector<std::vector<uint8_t>> MongoDB::GetNextXids(int batch_size) {
  try {
    // Atomically increment the counter and get the starting value
    auto opts = mongocxx::options::find_one_and_update{};
    opts.return_document(mongocxx::options::return_document::k_after);
    opts.upsert(true);

    auto result = counters_.find_one_and_update(
        document{} << "_id" << counter_key_ << finalize,
        document{} << "$inc" << open_document << "seq" << batch_size
                   << close_document << finalize,
        opts);

    if (!result) {
      throw std::runtime_error("Failed to increment counter");
    }

    auto doc_view = result->view();
    int64_t last_id = doc_view["seq"].get_int64();
    int64_t first_id = last_id - batch_size + 1;

    // Generate xids as 8-byte big-endian values
    std::vector<std::vector<uint8_t>> xids;
    xids.reserve(batch_size);

    for (int64_t i = first_id; i <= last_id; ++i) {
      std::vector<uint8_t> xid(8);
      // Big-endian encoding
      xid[0] = static_cast<uint8_t>((i >> 56) & 0xFF);
      xid[1] = static_cast<uint8_t>((i >> 48) & 0xFF);
      xid[2] = static_cast<uint8_t>((i >> 40) & 0xFF);
      xid[3] = static_cast<uint8_t>((i >> 32) & 0xFF);
      xid[4] = static_cast<uint8_t>((i >> 24) & 0xFF);
      xid[5] = static_cast<uint8_t>((i >> 16) & 0xFF);
      xid[6] = static_cast<uint8_t>((i >> 8) & 0xFF);
      xid[7] = static_cast<uint8_t>(i & 0xFF);
      xids.push_back(std::move(xid));
    }

    return xids;
  } catch (const std::exception& e) {
    throw std::runtime_error(std::string("Failed to get next xids: ") + e.what());
  }
}

std::pair<bool, std::vector<std::string>> MongoDB::Put(
    const std::vector<std::string>& keys,
    const std::vector<std::vector<uint8_t>>& values,
    const std::vector<std::vector<uint8_t>>* xids) {
  try {
    if (keys.size() != values.size()) {
      throw std::invalid_argument("keys and values must have the same length");
    }

    if (xids && xids->size() != keys.size()) {
      throw std::invalid_argument("xids and keys must have the same length");
    }

    size_t n = keys.size();
    std::vector<std::vector<uint8_t>> auto_xids;

    // Determine which xids to use
    const std::vector<std::vector<uint8_t>>* xids_to_use = xids;
    if (!xids) {
      auto_xids = GetNextXids(n);
      xids_to_use = &auto_xids;
    }

    std::vector<std::string> inserted_ids;

    // Perform individual upserts for each key-value pair
    for (size_t i = 0; i < n; ++i) {
      // Build update document
      auto update_content = document{};
      update_content << "key" << keys[i];

      // Store value as binary data
      update_content << "value" << bsoncxx::types::b_binary{
          bsoncxx::binary_sub_type::k_uuid,
          static_cast<uint32_t>(values[i].size()),
          reinterpret_cast<const uint8_t*>(values[i].data())
      };

      // Store xid as binary data
      update_content << "xid" << bsoncxx::types::b_binary{
          bsoncxx::binary_sub_type::k_uuid,
          static_cast<uint32_t>((*xids_to_use)[i].size()),
          reinterpret_cast<const uint8_t*>((*xids_to_use)[i].data())
      };

      // Perform upsert
      auto filter = document{} << "key" << keys[i] << finalize;
      auto update = document{} << "$set" << update_content << finalize;

      mongocxx::options::update opts{};
      opts.upsert(true);

      collection_.update_one(filter.view(), update.view(), opts);
      inserted_ids.push_back(keys[i]);
    }

    return {true, inserted_ids};
  } catch (const std::exception& e) {
    return {false, {}};
  }
}

std::vector<uint8_t> MongoDB::ExtractValueFromDoc(
    const bsoncxx::document::view& doc) {
  try {
    auto value_element = doc["value"];
    if (value_element && value_element.type() == bsoncxx::type::k_binary) {
      auto binary = value_element.get_binary();
      const uint8_t* data = binary.bytes;
      uint32_t size = binary.size;
      return std::vector<uint8_t>(data, data + size);
    }
  } catch (...) {
  }
  return std::vector<uint8_t>();
}

std::vector<uint8_t> MongoDB::ExtractXidFromDoc(
    const bsoncxx::document::view& doc) {
  try {
    auto xid_element = doc["xid"];
    if (xid_element && xid_element.type() == bsoncxx::type::k_binary) {
      auto binary = xid_element.get_binary();
      const uint8_t* data = binary.bytes;
      uint32_t size = binary.size;
      return std::vector<uint8_t>(data, data + size);
    }
  } catch (...) {
  }
  return std::vector<uint8_t>();
}

std::pair<std::vector<std::vector<uint8_t>>, std::vector<std::vector<uint8_t>>>
MongoDB::GetByKeys(const std::vector<std::string>& keys) {
  std::vector<std::vector<uint8_t>> values;
  std::vector<std::vector<uint8_t>> xids_result;

  if (keys.empty()) {
    return {values, xids_result};
  }

  try {
    // Build $in query for multiple keys
    auto in_array = array{};
    for (const auto& key : keys) {
      in_array << key;
    }

    auto query = document{} << "key" << open_document << "$in" << in_array
                            << close_document << finalize;

    // Execute query and build result maps
    auto cursor = collection_.find(query.view());

    std::map<std::string, std::pair<std::vector<uint8_t>, std::vector<uint8_t>>> 
        key_to_data;

    for (auto&& doc : cursor) {
      auto key_element = doc["key"];
      if (key_element && key_element.type() == bsoncxx::type::k_string) {
        std::string key{key_element.get_string().value};
        auto value = ExtractValueFromDoc(doc);
        auto xid = ExtractXidFromDoc(doc);
        key_to_data[key] = std::make_pair(value, xid);
      }
    }

    // Build result vectors preserving input key order
    for (const auto& key : keys) {
      auto it = key_to_data.find(key);
      if (it != key_to_data.end()) {
        values.push_back(it->second.first);
        xids_result.push_back(it->second.second);
      } else {
        values.push_back(std::vector<uint8_t>());
        xids_result.push_back(std::vector<uint8_t>());
      }
    }

    return {values, xids_result};
  } catch (const std::exception& e) {
    // On error, return empty results for all keys
    values.assign(keys.size(), std::vector<uint8_t>());
    xids_result.assign(keys.size(), std::vector<uint8_t>());
    return {values, xids_result};
  }
}

std::vector<std::vector<uint8_t>> MongoDB::GetByIds(
    const std::vector<std::vector<uint8_t>>& xids) {
  std::vector<std::vector<uint8_t>> results;

  if (xids.empty()) {
    return results;
  }

  try {
    // Build $in query for multiple xids
    auto in_array = array{};
    for (const auto& xid : xids) {
      in_array << bsoncxx::types::b_binary{
          bsoncxx::binary_sub_type::k_uuid,
          static_cast<uint32_t>(xid.size()),
          reinterpret_cast<const uint8_t*>(xid.data())
      };
    }

    auto query = document{} << "xid" << open_document << "$in" << in_array
                            << close_document << finalize;

    // Execute query and build result map
    auto cursor = collection_.find(query.view());

    std::map<std::vector<uint8_t>, std::vector<uint8_t>> xid_to_value;

    for (auto&& doc : cursor) {
      auto xid = ExtractXidFromDoc(doc);
      if (!xid.empty()) {
        auto value = ExtractValueFromDoc(doc);
        xid_to_value[xid] = value;
      }
    }

    // Build result vector preserving input xid order
    for (const auto& xid : xids) {
      auto it = xid_to_value.find(xid);
      if (it != xid_to_value.end()) {
        results.push_back(it->second);
      } else {
        results.push_back(std::vector<uint8_t>());
      }
    }

    return results;
  } catch (const std::exception& e) {
    // On error, return empty results for all xids
    results.assign(xids.size(), std::vector<uint8_t>());
    return results;
  }
}

bool MongoDB::Delete(const std::string& key) {
  try {
    auto result = collection_.delete_one(
        document{} << "key" << key << finalize);
    return result->deleted_count() > 0;
  } catch (...) {
    return false;
  }
}

}  // namespace hakes
